File: 2023/01/01.py

import re
def preprocessing(input):
    document = list()
    for line in input.splitlines():
        matches = re.finditer(r'(?=([0-9]|one|two|three|four|five|six|seven|eight|nine))', line)
        numbers = [match.group(1) for match in matches]
        document.append(numbers)
    return document

def solver(document):
    allowed = {str(n): n for n in range(1, 10)}
    yield calibrate(document.copy(), allowed)
    
    allowed.update({
        "one":   1,
        "two":   2,
        "three": 3,
        "four":  4,
        "five":  5, 
        "six":   6,
        "seven": 7,
        "eight": 8, 
        "nine":  9
    })
    yield calibrate(document, allowed)

def calibrate(document, allowed):
    calibration_values = 0
    for digits in document:
        l = 0
        r = len(digits) - 1
        while (a:= digits[l]) not in allowed.keys(): l += 1
        while (b:= digits[r]) not in allowed.keys(): r -= 1
        calibration_values += allowed[a] * 10 + allowed[b]
    return calibration_values

File: 2023/01/01.rs
aoc::main!();

fn preprocessing(input: &str) -> Vec<Vec<usize>> {
    input
    .lines()
    .map(|line| 
        AhoCorasick::new(&["1","2","3","4","5","6","7","8","9","_","one","two","three","four","five","six","seven","eight","nine"])
        .unwrap()
        .find_overlapping_iter(line)
        .map(|mat| mat.pattern().as_usize() + 1)
        .collect())
    .collect()
    }

fn part_1(document: Vec<Vec<usize>>) -> usize {
    document
    .iter()
    .map(|digits| 
        digits
        .iter()
        .filter(|&d| *d < 10))
    .map(|mut digits| 
        10 * digits.clone().nth(0).unwrap() 
        + digits.nth_back(0).unwrap())
    .sum()
}

fn part_2(document: Vec<Vec<usize>>) -> usize {
    document
    .iter()
    .map(|digits|
        digits
        .iter()
        .map(|d| d % 10))
    .map(|mut digits| 
        10 * digits.clone().nth(0).unwrap() 
        + digits.nth_back(0).unwrap())
    .sum()
}